### PR TITLE
Decaying histograms

### PR DESCRIPTION
**Pull request by kgrygiel**

### CODE REVIEW

File: vertical-pod-autoscaler/recommender/util/decaying_histogram.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+// A histogram that gives newer samples a higher weight than the old samples,
+// gradually "forgetting" the past samples. The weight of each sample is
+// multiplied by the factor of 2^((sampleTime - decayStart) / halfLife).
+// This means that the sample loses half of its weight ("importance") after
+// each halfLife period.
+// Since only relative (and not absolute) weights of samples matter, the
+// decayStart can be shifted at any time, which is equivalent to multiplying all
+// weights by a constant. In fact this happens whenever the exponents become too
+// large, to stay in the range of the floating point arithmentics.
```
Comment (bskiba) [inline vertical-pod-autoscaler/recommender/util/decaying_histogram.go]: nit: s/arithmentics/arithmetics/

**Comment by bskiba** [comment]
@mwielgus do you want to take a look too?

File: vertical-pod-autoscaler/recommender/util/histogram.go [hunk 2]
```diff
@@ -157,25 +161,40 @@ func (h *histogram) IsEmpty() bool {
...
 	return strings.Join(lines, "\n")
 }
 
-// HistogramsEqual is a helper function for comparing 2 histograms.
-func HistogramsEqual(histogram1 Histogram, histogram2 Histogram) bool {
-	h1 := (histogram1).(*histogram)
-	h2 := (histogram2).(*histogram)
-	if h1.options != h2.options || h1.minBucket != h2.minBucket || h1.maxBucket != h2.maxBucket {
+func (h *histogram) Equals(other Histogram) bool {
+	h2, typesMatch := (other).(*histogram)
```
Comment (tkulczynski) [inline vertical-pod-autoscaler/recommender/util/histogram.go]: The object for type assertion is not typically enclosed in paranthesis: https://tour.golang.org/methods/15.

File: vertical-pod-autoscaler/recommender/util/decaying_histogram.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+func (h *decayingHistogram) shiftDecayStart(newDecayStart time.Time) {
+	// Make sure the decay start is an integer multiple of halfLife.
+	newDecayStart = newDecayStart.Round(h.halfLife)
+	exponent := round(float64(h.decayStart.Sub(newDecayStart)) / float64(h.halfLife))
+	for bucket := h.histogram.minBucket; bucket <= h.histogram.maxBucket; bucket++ {
+		// Bucket weight *= 2^exponent.
+		h.histogram.bucketWeight[bucket] = math.Ldexp(h.histogram.bucketWeight[bucket], exponent)
+	}
+	h.histogram.totalWeight = math.Ldexp(h.histogram.totalWeight, exponent)
```
Comment (tkulczynski) [inline vertical-pod-autoscaler/recommender/util/decaying_histogram.go]: I would probably make this a method in histogram that multiplies all buckets and totalWeight by the same constant. I might be thinking c++ too much, where things are private between classes.

File: vertical-pod-autoscaler/recommender/util/histogram.go [hunk 4]
```diff
@@ -157,25 +161,40 @@ func (h *histogram) IsEmpty() bool {
...
 			return false
 		}
 	}
 	return true
 }
+
+// Adjusts the value of minBucket and maxBucket after any operation that
+// decreases weights.
+func (h *histogram) updateMinAndMaxBucket() {
+	epsilon := (*h.options).Epsilon()
```
[the remark this answers is not in the document]
Comment (kgrygiel) [reply vertical-pod-autoscaler/recommender/util/histogram.go]: Done.

File: vertical-pod-autoscaler/recommender/util/histogram.go (hunk 2 above)
Comment (kgrygiel) [reply vertical-pod-autoscaler/recommender/util/histogram.go]: Done.

File: vertical-pod-autoscaler/recommender/util/decaying_histogram.go (hunk 3 above)
Comment (kgrygiel) [reply vertical-pod-autoscaler/recommender/util/decaying_histogram.go]: Done.

**Pull request merged**